src/optimizer/plan_nodes/mod.rs

```diff
@@ -87,13 +88,50 @@ pub use physical_simple_agg::*;
 pub use physical_table_scan::*;
 pub use physical_values::*;
 
+/// The upcast trait for `PlanNode`.
+pub trait IntoPlanRef {
+    fn into_plan_ref(self) -> PlanRef;
+    fn clone_as_plan_ref(&self) -> PlanRef;
+}
 /// The common trait over all plan nodes.
 pub trait PlanNode:
-    WithPlanNodeType + PlanTreeNode + Debug + Display + Downcast + erased_serde::Serialize + Send + Sync
+    WithPlanNodeType
+    + IntoPlanRef
+    + PlanTreeNode
+    + Debug
+    + Display
+    + Downcast
+    + erased_serde::Serialize
+    + Send
+    + Sync
 {
     fn out_types(&self) -> Vec<DataType> {
         vec![]
     }
+    /// transform the plan node to only output the required columns ordered by index number, only
+    /// logical plan node will use it, though all plan node impl it.
+    fn prune_col(&self, required_cols: BitSet) -> PlanRef {
+        let input_types = self.out_types();
+        let mut need_prune = false;
+        for i in 0..input_types.len() {
+            if !required_cols.contains(i) {
+                need_prune = true;
+            }
+        }
+        if !need_prune {
+            return self.clone_as_plan_ref();
+        }
+        let exprs = required_cols
+            .iter()
+            .map(|index| {
+                BoundExpr::InputRef(BoundInputRef {
+                    index,
+                    return_type: input_types[index].clone(),
+                })
+            })
+            .collect();
+        LogicalProjection::new(exprs, self.clone_as_plan_ref()).into_plan_ref()
```

Comment:
Simply doing projection won't save I/Os. Will we have plan rules to merge `LogicalProjection` with `TableScan`, and merge `LogicalProjection` with `LogicalProjection`?

Reply:
Here we simply doing projection to maintain the correct behavior of the `fn prune_col`. in future, every PlanNode should implement its own `prune_col` method, which can call its inputs' `prune_col` recursively. in the end, when call the `prune_col` of the TableScan Node, it will remove the columns not required.